poc/flp_generic.py

```diff
@@ -888,3 +918,88 @@ def test_vec_set_type_param(self, test_vec):
         test_vec['bits'] = self.bits
         test_vec['chunk_length'] = self.chunk_length
         return ['length', 'bits', 'chunk_length']
+
+
+# TODO(issue #306) Replace `Sum` with this type.
+class SumOfRangeCheckedInputs(Valid):
+    # Associated types
+    Measurement = int  # `range(self.max_measurement+1)`
+    AggResult = int
+    Field = field.Field64
+
+    # Associated parameters
+    GADGETS = [Range2()]
+    JOINT_RAND_LEN = 0
+    OUTPUT_LEN = 1
+
+    def __init__(self, max_measurement: int):
+        """
+        Similar to `Sum` but with an arbitrary bound.
+
+        The circuit checks that the measurement is in
+        `range(max_measurement+1)`. This is accomplished by encoding the
+        measurement in a way that ensures it is in range, then comparing the
+        reported measurement to the range checked measurement.
+
+        Let
+
+        - `bits = max_measurement.bit_length()`
+        - `offset = 2**bits - 1 - max_measurement`
+
+        The range checked measurement is the bit-encoding of `offset` plus the
+        measurement. Observe that only measurements in at most
+        `max_measurement` can be encoded with `bits` bits.
+
+        To do the range check, the circuit first checks that each
+        entry of this bit vector is a one or a zero. It then decodes
+        it and subtracts it from `offset` plus the reported value.
+        Since the range checked measurement is in the correct range,
+        equality implies that the reported measurement is as well.
+        """
+        self.bits = max_measurement.bit_length()
+        self.offset = self.Field(2**self.bits - 1 - max_measurement)
+        self.max_measurement = max_measurement
+
+        if 2**self.bits >= self.Field.MODULUS:
+            raise ValueError('bound exceeds field modulus')
+
+        self.GADGET_CALLS = [2*self.bits]
+        self.MEAS_LEN = 2*self.bits
+        self.EVAL_OUTPUT_LEN = 2*self.bits + 1
+
+    def eval(self, meas, joint_rand, num_shares):
+        self.check_valid_eval(meas, joint_rand)
+        shares_inv = self.Field(num_shares).inv()
+
+        out = []
+        for b in meas:
+            out.append(self.GADGETS[0].eval(self.Field, [b]))
+
+        range_check = self.offset*shares_inv + \
+            self.Field.decode_from_bit_vector(meas[:self.bits]) - \
+            self.Field.decode_from_bit_vector(meas[self.bits:])
+        out.append(range_check)
+        return out
+
+    def encode(self, measurement):
+        x = self.Field(measurement)
+        encoded = []
+        encoded += self.Field.encode_into_bit_vector(
+            x.as_unsigned(),
+            self.bits
+        )
+        encoded += self.Field.encode_into_bit_vector(
+            (x + self.offset).as_unsigned(),
+            self.bits
+        )
+        return encoded
```

Comment:
This could be simplified a little now
```suggestion
        encoded = []
        encoded += self.Field.encode_into_bit_vector(
            measurement,
            self.bits
        )
        encoded += self.Field.encode_into_bit_vector(
            measurement + self.offset.as_unsigned(),
            self.bits
        )
        return encoded
```